[pytorch][codegen] add autograd data model

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48249 [pytorch][codegen] add autograd data model**

Introduced autograd related data models at tools.codegen.api.autograd.

Migrated load_derivatives.py to produce the new data models from derivatives.yaml.
It has clean mypy-strict result.

Changed both gen_autograd_functions.py and gen_variable_type.py to consume
the new data model.

Added type annotations to gen_autograd_functions.py - it has clean mypy-strict
result except for the .gen_autograd import (so haven't added it to the strict
config in this PR).

To limit the scope of the PR, gen_variable_type.py is not refactored, and the
main structure of load_derivatives.py / gen_autograd_functions.py is kept. We
only make necessary changes to make it work.

Confirmed byte-for-byte compatible with the old codegen:

```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D25086561](https://our.internmc.facebook.com/intern/diff/D25086561)

ljk53 committed:
[pytorch][codegen] add autograd data model

Introduced autograd related data models at tools.codegen.api.autograd.

Migrated load_derivatives.py to produce the new data models from derivatives.yaml.
It has clean mypy-strict result.

Changed both gen_autograd_functions.py and gen_variable_type.py to consume
the new data model.

Added type annotations to gen_autograd_functions.py - it has clean mypy-strict
result except for the .gen_autograd import (so haven't added it to the strict
config in this PR).

To limit the scope of the PR, gen_variable_type.py is not refactored, and the
main structure of load_derivatives.py / gen_autograd_functions.py is kept. We
only make necessary changes to make it work.

Confirmed byte-for-byte compatible with the old codegen:

```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] add autograd data model"


Introduced autograd related data models at tools.codegen.api.autograd.

Migrated load_derivatives.py to produce the new data models from derivatives.yaml.
It has clean mypy-strict result.

Changed both gen_autograd_functions.py and gen_variable_type.py to consume
the new data model.

Added type annotations to gen_autograd_functions.py - it has clean mypy-strict
result except for the .gen_autograd import (so haven't added it to the strict
config in this PR).

To limit the scope of the PR, gen_variable_type.py is not refactored, and the
main structure of load_derivatives.py / gen_autograd_functions.py is kept. We
only make necessary changes to make it work.

Confirmed byte-for-byte compatible with the old codegen:

```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] add autograd data model"


Introduced autograd related data models at tools.codegen.api.autograd.

Migrated load_derivatives.py to produce the new data models from derivatives.yaml.
It has clean mypy-strict result.

Changed both gen_autograd_functions.py and gen_variable_type.py to consume
the new data model.

Added type annotations to gen_autograd_functions.py - it has clean mypy-strict
result except for the .gen_autograd import (so haven't added it to the strict
config in this PR).

To limit the scope of the PR, gen_variable_type.py is not refactored, and the
main structure of load_derivatives.py / gen_autograd_functions.py is kept. We
only make necessary changes to make it work.

Confirmed byte-for-byte compatible with the old codegen:

```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D25086561](https://our.internmc.facebook.com/intern/diff/D25086561)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] add autograd data model"


Introduced autograd related data models at tools.codegen.api.autograd.

Migrated load_derivatives.py to produce the new data models from derivatives.yaml.
It has clean mypy-strict result.

Changed both gen_autograd_functions.py and gen_variable_type.py to consume
the new data model.

Added type annotations to gen_autograd_functions.py - it has clean mypy-strict
result except for the .gen_autograd import (so haven't added it to the strict
config in this PR).

To limit the scope of the PR, gen_variable_type.py is not refactored, and the
main structure of load_derivatives.py / gen_autograd_functions.py is kept. We
only make necessary changes to make it work.

Confirmed byte-for-byte compatible with the old codegen:

```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D25086561](https://our.internmc.facebook.com/intern/diff/D25086561)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] add autograd data model"


Introduced autograd related data models at tools.codegen.api.autograd.

Migrated load_derivatives.py to produce the new data models from derivatives.yaml.
It has clean mypy-strict result.

Changed both gen_autograd_functions.py and gen_variable_type.py to consume
the new data model.

Added type annotations to gen_autograd_functions.py - it has clean mypy-strict
result except for the .gen_autograd import (so haven't added it to the strict
config in this PR).

To limit the scope of the PR, gen_variable_type.py is not refactored, and the
main structure of load_derivatives.py / gen_autograd_functions.py is kept. We
only make necessary changes to make it work.

Confirmed byte-for-byte compatible with the old codegen:

```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D25086561](https://our.internmc.facebook.com/intern/diff/D25086561)

[ghstack-poisoned]

ezyang commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -84,142 +90,143 @@
+    compute_index_ranges: List[str] = []
+
+    for arg in info.args_with_derivatives:
+        if arg.type == 'TensorList':
```
bare string test? :o

ezyang commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -84,142 +90,143 @@
+
+    def save_var(var: Variable, is_output: bool) -> None:
+        name = var.name
+        if var.type == 'Tensor' or var.type == 'c10::optional<Tensor>' or var.type == 'c10::optional<Tensor>&' or \
```
This is less refactor than I was hoping for, but this might actually be better since you're about to conflict with @albanD, the review has been pretty straightforward since things line up.

albanD replied:
The addition of forward AD might impose some more constraints. So I think we should wait on more refactor until it is there yes.

ezyang commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -84,142 +90,143 @@
+    def emit_derivative(
+        derivative: Derivative,
+        args_with_derivatives: Sequence[CppArgument],
+    ) -> Tuple[bool, str]:
```
These types are so awesome

ezyang replied:
No idea what the bool return is though 🤣

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -899,3 +912,63 @@ def dispatch_strategy(declaration):
+    types = ', '.join(python.argument_type_str(a.argument.type, simple_type=True)
+                      if isinstance(a.argument, Argument) else 'TensorOptions'
+                      for a in args)
+    return f'{cpp.name(f.func)}({types})'
```
cpp.name hmmmmmm

ezyang commented on tools/autograd/load_derivatives.py:
```diff
@@ -334,90 +300,75 @@ def stride_expr(name):
+                ))
                 if 'res' in info:
-                    return info['res'](name)
+                    replacement: str = info['res'](name)
```
Odd that you need a type annotation here

ljk53 replied:
`info['res']` is a lambda - mypy simply takes it as returning Any.

ezyang commented:
Most of the changes here are pretty mild; the most major change is `load_derivatives.py` which had some surgery to avoid mutating fields inplace.

ezyang reviewed (approved):
Looks great, thank you!

albanD commented on tools/codegen/model.py:
```diff
@@ -527,7 +527,6 @@ def kind(self) -> SchemaKind:
         else:
             return SchemaKind.functional
 
-    # WARNING: This method is not currently tested in any meaningful way
```
It is properly tested now? :D

ljk53 replied:
I think @ezyang created this method for the autograd codegen work, but it was not integrated until this PR.
We expect it to generate the same canonical signature string for inplace/out/functional versions of the same function (so we can group them together). It seems to work so far :)

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -0,0 +1,76 @@
+from tools.codegen.api.types import *
+from tools.codegen.model import *
+
+# Represents a saved variable invovled in backward calculation.
```
nit: involved

ljk53 replied:
done

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -0,0 +1,76 @@
+# Note that it can be a derived property of an input argument, e.g.:
+# we could save `other.scalar_type()` instead of the entire `other` tensor.
+@dataclass(frozen=True)
+class Variable:
```
Maybe `SavedAttribute`

ljk53 replied:
renamed

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -0,0 +1,76 @@
+@dataclass(frozen=True)
+class Derivative:
+    # The formula string (legit C++ expression).
+    # Note that expressions against input arguments are replaced with the
```
nit: it is not super clear if this replacement is already done in this string or if it needs to be done before using it?

ljk53 replied:
'are' -> 'have been' and updated the wording below

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -0,0 +1,76 @@
+    # The base name read from derivatives.yaml.
+    name: str
+
+    # The matching native function.
```
nit: can we add a comment there to say if it is the exact same overload or the base version?

ljk53 replied:
Added inline comments:
```
    # There can be multiple NativeFunction having the same base name:
    #  - different overloads with different types of input arguments;
    #  - in-place/out/functional variants of the same function;
    #
    # We first use the schema string (under the 'name' key) in derivatives.yaml
    # to find the NativeFunction having the same schema string.
    # Then we find the in-place/out/functional variants of the matching function.
    # Among these variants, we choose the one having the same name as the
    # derivatives.yaml entry. If there is no exact match, then we choose the
    # in-place variant, if any.
    # TODO: maybe the logic to search for all variants is no longer necessary?
```

After taking a closer look, I actually feel the `canonical_function()` doesn't seem to be necessary:
In the first step, we already find the function with the exact same function schema string - there
shouldn't be a better match among in-place/out/functional variants.

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -0,0 +1,76 @@
+    # Names of arguments whose derivative formula is 'non_differentiable'.
+    non_differentiable_arg_names: Sequence[str]
+
+    # Raw data read from derivatives.yaml.
```
Is this always set? Or can it be empty when not specified in derivatives.yaml?

ljk53 replied:
The type annotation has Optional in it: `Optional[List[bool]]`

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -1,115 +1,115 @@
+
+    # What's the difference between function schema v.s. signature?
+    # function schema is the complete declaration including mutability annotation / default value and etc.
+    # signature is the canonical schema for a group of functions (in/out/functional variants)
```
What does the "in" variant means here?

ljk53 replied:
Changed it to 'in-place' - referencing the terms used in model.py

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -1,115 +1,115 @@
+    # 'XyzBackward' instead of 'XyzBackward0' or unconditionally append '0' to singletons.
+    op_names = create_op_names(infos)
+    return [
+        DifferentiabilityInfo(
```
This is just to add the op_name to each DifferentiabilityInfo but because they're unmutable, we need to recreate tall of them right?

ljk53 replied:
correct

albanD commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -4,11 +4,17 @@
+
+from typing import List, Sequence, Tuple, Optional
+
+from tools.codegen.api.autograd import *
```
nit: you always import the codegen part with `*` what is the reason here? it makes quite a few names below hard to track down where they come from :/

ljk53 replied:
I can add explicit imports - @ezyang shall we make it a convention?

albanD reviewed (commented):
Looks good!
Mostly naming/comments

ljk53 committed:
Update on "[pytorch][codegen] add autograd data model"


Introduced autograd related data models at tools.codegen.api.autograd.

Migrated load_derivatives.py to produce the new data models from derivatives.yaml.
It has clean mypy-strict result.

Changed both gen_autograd_functions.py and gen_variable_type.py to consume
the new data model.

Added type annotations to gen_autograd_functions.py - it has clean mypy-strict
result except for the .gen_autograd import (so haven't added it to the strict
config in this PR).

To limit the scope of the PR, gen_variable_type.py is not refactored, and the
main structure of load_derivatives.py / gen_autograd_functions.py is kept. We
only make necessary changes to make it work.

Confirmed byte-for-byte compatible with the old codegen:

```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D25086561](https://our.internmc.facebook.com/intern/diff/D25086561)

[ghstack-poisoned]
